File: bestia-client/src/Bnet/Message/Master/InvestSkillPointCMSG.cs
using Bnet;
using Godot;

namespace BestiaBehemothClient.Bnet.Message.Master
{
  /// <summary>
  /// Message to spend one or more of the master's available skill points across one or more
  /// skill tree nodes in a single batch request. Each entry's "attack_id" key is the skill's
  /// catalog id (matches skills.yml / SkillListSMSG.skillId), "amount" is how many additional
  /// levels to invest into it.
  /// </summary>
  public partial class InvestSkillPointCMSG : ICMSG
  {
    [Export] public Godot.Collections.Array<Godot.Collections.Dictionary> InvestedPoints { get; set; } = new Godot.Collections.Array<Godot.Collections.Dictionary>();

    public InvestSkillPointCMSG()
    {
    }

    public override Envelope ToEnvelope()
    {
      var investSkillPointCmsg = new global::Bnet.InvestSkillPointCMSG();
      foreach (var entry in InvestedPoints)
      {
        investSkillPointCmsg.InvestedPoints.Add(new global::Bnet.InvestedSkillPoint
        {
          AttackId = (ulong)entry["attack_id"].AsInt64(),
          Amount = (uint)entry["amount"].AsInt32()
        });
      }

      return new Envelope
      {
        InvestSkillPoint = investSkillPointCmsg
      };
    }
  }
}

File: zone-server/src/main/kotlin/net/bestia/zone/account/master/skill/InvestSkillPointCMSG.kt
package net.bestia.zone.account.master.skill

import net.bestia.bnet.proto.InvestSkillPointCmsgProto
import net.bestia.zone.message.CMSG

data class InvestSkillPointCMSG(
  override val playerId: Long,
  val investedPoints: List<InvestedSkillPoint>
) : CMSG {

  data class InvestedSkillPoint(
    val attackId: Long,
    val amount: Int
  )

  companion object {
    fun fromBnet(
      accountId: Long,
      investSkillPoint: InvestSkillPointCmsgProto.InvestSkillPointCMSG
    ): InvestSkillPointCMSG {
      return InvestSkillPointCMSG(
        playerId = accountId,
        investedPoints = investSkillPoint.investedPointsList.map {
          InvestedSkillPoint(attackId = it.attackId, amount = it.amount)
        }
      )
    }
  }
}
